### PR TITLE
Sean category api

### PR DESCRIPTION
**Pull request by w-disheng**

### CODE REVIEW

**Comment by francisli** [comment]
@w-disheng there are conflicts with main branch that need to be resolved in this branch now...

Is it ready for review now, or are you still working on it?

**Comment by w-disheng** [comment]
> 
> 
> @w-disheng there are conflicts with main branch that need to be resolved in this branch now...
> 
> Is it ready for review now, or are you still working on it?

Sorry Francis, Was kinda busy during weekend. My friend went back to China. Yes, now is ready.

**Comment by francisli** [comment]
I see you just did a merge from main... do be careful to RUN and TEST your code after a merge, especially after a conflict, as I've just pulled down the branch and it is broken... will patch and push up...

File: client/src/components/category/Categories.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  return (
+    <Switch>
+      <Route exact path={path}>
+        <CategoriesList />
+      </Route>
+      <Route path={`${path}/new`}>
+        <CategoryForm />
+      </Route>
+      <Route path={`${path}/:id/edit`}>
+        <CategoriesList />
```
Comment (francisli) [inline client/src/components/category/Categories.js]: This is incorrect. The edit route should display the CategoryForm component.

File: client/src/components/category/CategoryForm.js [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    }
+  }, [id]);
+
+  const onChange = (e) => {
+    const { name, value } = e.target;
+    const newCategory = { ...category };
+    if (value === "") {
+      newCategory[name] = value;
+    } else {
+      newCategory[name] = parseFloat(value);
```
Comment (francisli) [inline client/src/components/category/CategoryForm.js]: This code is causing editing the name to fail, because once you start entering text, it trys to parse it into a float.

I'm not even sure this is entirely necessary- the DB should be able to convert from string to decimal as needed...

File: routes/api/categories.js [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+router.get("/", async function (req, res) {
+  const row = await models.Category.findAll();
+  res.json(row);
+});
+
+router.post("/", interceptors.requireLogin, async function (req, res) {
+  const user = req.user;
+  try {
+    let category = await user.getCategory();
```
Comment (francisli) [inline routes/api/categories.js]: Copying this code from the api/budgets is incorrect. The reason is that User/Budget has a ONE-TO-ONE relationship, so you never can create more than one Budget, so the code checks if that Budget exists already.

However, User/Category is ONE-TO-MANY, so you need to allow for the creation of as many as the User wants/needs.

Also, the User/Category association has not been set up yet in the models, so this code fails.

File: client/src/components/category/CategoriesList.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        {categories.map((c) => (
+          <li>
+            <p>
+              <Link to={`/categories/${c.id}/edit`}>
+                {c.categoryName}, {c.limit}, {c.spend}
+              </Link>
+            </p>
+            <p>
+              <button
+                onClick={() => onDelete(s)}
```
Comment (francisli) [inline client/src/components/category/CategoriesList.js]: The variable in the map function is c, not s, so this code will fail.

File: client/src/components/category/CategoriesList.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+function CategoriesList() {
+  const [categories, setCategories] = useState([]);
+
+  useEffect(() => {
+    Api.categories.index().then((response) => setCategories(response.data));
+  }, []);
+
+  const onDelete = (category) => {
+    if (window.confirm(`Are you sure you want to delete ${category.name}?`)) {
```
Comment (francisli) [inline client/src/components/category/CategoriesList.js]: Isn't the attribute called categoryName, not just name?

**Review by francisli — CHANGES_REQUESTED** [review]
@w-disheng I'm a bit disappointed that you clearly did not run/test this code, for me to find these types of bugs in it...!

File: migrations/20210519031304-add-category-references.js [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+     */
+    await queryInterface.addConstraint("Categories", {
+      fields: ["UserId"],
+      type: "FOREIGN KEY",
+      references: {
+        table: "Users",
+        field: "id",
+      },
+    });
+    await queryInterface.addConstraint("Categories", {
```
Comment (francisli) [inline migrations/20210519031304-add-category-references.js]: It is incorrect to add a UNIQUE constraint, because that would make User/Category ONE-TO-ONE

File: models/category.js [hunk 7]
```diff
@@ -11,16 +9,19 @@ module.exports = (sequelize, DataTypes) => {
      */
     static associate(models) {
       // define association here
+      Category.belongsTo(models.User);
```
Comment (francisli) [inline models/category.js:12]: You should also define the corresponding association in models/user.js.

File: routes/api/categories.js (hunk 3 above)
Comment (francisli) [reply routes/api/categories.js]: Remove the check for an existing category, and just create a brand new Category as with earlier api route examples...

**Comment by w-disheng** [comment]
> @w-disheng I'm a bit disappointed that you clearly did not run/test this code, for me to find these types of bugs in it...!

Will fix it now, I  am sorry. Docker only work on my Mac Air been trying many way still not ran.   And I have been using Window for works since my mac slow and dc from internet all the time.  I should not be lazy to test before I ask for review. Will be more careful next time.

**Comment by francisli** [comment]
@w-disheng do you know why Docker isn't running on your Windows PC?

File: client/src/components/category/Categories.js [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+  const { path } = useRouteMatch();
+  return (
+    <Switch>
+      <Route exact path={path}>
+        <CategoriesList />
+      </Route>
+      <Route path={`${path}/new`}>
+        <CategoryForm />
+      </Route>
+      <Route path={`${path}/:id/form`}>
```
Comment (francisli) [inline client/src/components/category/Categories.js]: Why did you rename the route from ../edit to .../form? That breaks the links in the list page...

**Review by francisli — CHANGES_REQUESTED** [review]

**Review by francisli — APPROVED** [review]
Looks like everything is working now!
